openehr-rm/src/main/java/com/nedap/archie/rm/ehr/Ehr.java

```diff
@@ -171,4 +173,49 @@ public boolean equals(Object o) {
     public int hashCode() {
         return Objects.hash(systemId, ehrId, contributions, ehrStatus, ehrAccess, compositions, directory, timeCreated, folders);
     }
+
+    @Invariant("Contributions valid")
+    public boolean contributionsValid() {
+        return InvariantUtil.objectRefTypeEquals(contributions, "CONTRIBUTION");
+    }
+
+    @Invariant("Ehr_access_valid")
+    public boolean ehrAccessValid() {
+        return InvariantUtil.objectRefTypeEquals(ehrAccess, "VERSIONED_EHR_ACCESS");
+
+    }
+
+    @Invariant("Ehr_status_valid")
+    public boolean ehrStatusValid() {
+        return InvariantUtil.objectRefTypeEquals(ehrStatus, "VERSIONED_EHR_STATUS");
+
+    }
+
+    @Invariant("Compositions_valid")
+    public boolean compositionsValid() {
+        return InvariantUtil.objectRefTypeEquals(compositions, "VERSIONED_COMPOSITION");
+
+    }
+
+    @Invariant("Directory_valid")
+    public boolean directoryValid() {
+        return InvariantUtil.objectRefTypeEquals(directory, "VERSIONED_FOLDER");
+
+    }
+
+
+    @Invariant("Folderss_valid")
+    public boolean foldersValid() {
+        return InvariantUtil.objectRefTypeEquals(folders, "VERSIONED_FOLDER");
+
+    }
+
+    @Invariant("Directory_in_folders")
+    public boolean directoryInFolders() {
+        if (folders != null && directory != null) {
+            return folders.size() >= 1 && folders.get(0).equals(directory);
+        }
+        return true;
```

Comment:
Yes
<img width="972" alt="Screenshot 2021-03-09 at 12 51 04" src="https://user-images.githubusercontent.com/21126/110466543-2eb67c00-80d6-11eb-85fb-214acf81f043.png">

Reply:
What I meant was that this implementation allows `directory` to be `null` while `folders` it non-empty. But I think that should not be allowed because of:

> If folders is not Void, the directory attribute always contains a reference to the first member, for backward compatibility with pre-Release 1.1.0 systems.

from: https://specifications.openehr.org/releases/RM/latest/ehr.html#_folders)